scipy/linalg/tests/test_special_matrices.py

```diff
@@ -636,3 +639,40 @@ def test_fiedler_companion():
     fc = fiedler_companion([1., -16., 86., -176., 105.])
     assert_array_almost_equal(eigvals(fc),
                               np.array([7., 5., 3., 1.]))
+
+
+class TestConvolutionMatrix:
+    'test convolution_matrix vs. numpy.convolve for various parameters'
+
+    def create_vector(self, n, cpx):
+        'make a complex or real test vector of length n'
+        x = np.linspace(-2.5, 2.2, n)
+        if cpx:
+            x = x + 1j*np.linspace(-1.5, 3.1, n)
+        return x
+
+    def test_bad_first_arg(self):
+        # first arg must be a 1d array, otherwise ValueError
+        with assert_raises(ValueError):
```

Comment:
I find this pattern to be both more precise (ensures you catch the right thing) and avoid the need for a separate comment saying "why/what am I catching" (by putting that in the call itself):
```
with pytest.raises(ValueError, match='some suitable regex'):
```
So in your case you could do:
```
with pytest.raises(ValueError, match='expects a 1d array'):
```

Reply:
I like the idea of making the test more specific.  But is it generally understood that the text within an exception is subject to change between releases?  I don't think we've ever considered the text within an exception to be guaranteed to remain the same over time.  It is not unusual to change the message of an exception, perhaps to be more informative, or to fix grammar or style.  For example, in the future we might become more obsessive about using consistent abbreviations throughout the text presented to users, and we might decide that when `'one-dimensional'` is abbreviated, it should be `'1-d'`.  That would require changing the message in this exception, and that would break matching by the regex.  So using a regex match might be a bit fragile.

As long as we're all aware of this, and know that if we change the message we might have to change some tests that attempt to match that message, then there is no problem.  But I wouldn't want someone in the future to not improve an error message because an existing test has to match a specific string.


-----

An alternative method of making the tests more specific is to create a custom subclass of `ValueError` for each of the exceptions explicitly raised by the function, e.g. `ConvolutionMatrixInputNotOneDimensionalError`, `ConvolutionMatrixInvalidModeError`, etc., but I don't think anyone wants to go in that direction.

Reply:
> . But is it generally understood that the text within an exception is subject to change between releases? I don't think we've ever considered the text within an exception to be guaranteed to remain the same over time.

No it's not guaranteed to stay the same. But it's fairly trivial to update the test if the message changes or gets improved.

> So using a regex match might be a bit fragile.

It depends on how specific you make the regex. In my experience updating these (over the last couple years of operating under this pattern) is quite easy, and has caught numerous cases where a plain `ValueError` catch would have caught the *wrong* error, thereby making the test not catch what it meant to catch.

> As long as we're all aware of this, and know that if we change the message we might have to change some tests that attempt to match that message, then there is no problem

Agreed, this is what I have in mind

Reply:
Thanks, @larsoner, I just pushed an update.